### PR TITLE
Add query request and query response classes and tests

### PR DESCRIPTION
**Pull request by mziccard**
This PR adds classes for Query Requests and Query Responses:
- `QueryRequest` is used to run queries "on the fly" [Jobs: query](https://cloud.google.com/bigquery/docs/reference/v2/jobs/query)
- `QueryResponse` is used to both hold the result of queries run on the fly ([Jobs: query](https://cloud.google.com/bigquery/docs/reference/v2/jobs/query)) and to hold the result of query jobs ([Jobs: getQueryResults](https://cloud.google.com/bigquery/docs/reference/v2/jobs/getQueryResults))

**Remarks**
`QueryResponse` has no `fromPb` method as it is initialized through the builder inside the `BigQueryServiceImpl` class. This is so because:
- we need service and service options to initialize the field `Page<List<FieldValue>> rows`
- `QueryResponse` is created from 2 pb classes: [QueryResponse](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/model/QueryResponse.html) and [GetQueryResultsResponse](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/model/GetQueryResultsResponse.html)
  I preferred to handle the creation inside `BigQueryServiceImpl` rather than having a `fromPb` method with more parameters.

### CODE REVIEW

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+/**
+ * Google Cloud BigQuery Query Request. This class can be used to run a BigQuery SQL query and
+ * return results if the query completes within a specified timeout. The query results are saved to
+ * a temporary table that is deleted approximately 24 hours after the query is run. Query is run
+ * through a BigQuery Job whose identity can be accessed via {@link QueryResponse#job()}.
+ *
+ * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs/query">Query</a>
+ */
+public class QueryRequest implements Serializable {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: Looks like the two classes added are both serializable, they should be added to `SerializationTests`

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+import com.google.common.base.MoreObjects;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+/**
+ * Google Cloud BigQuery Query Request. This class can be used to run a BigQuery SQL query and
+ * return results if the query completes within a specified timeout. The query results are saved to
+ * a temporary table that is deleted approximately 24 hours after the query is run. Query is run
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: Nit: I think the sentence "Query is run..." would sound slightly better with a "The" at the beginning.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+     * to complete and read the results. The default value is 10000 milliseconds (10 seconds).
+     */
+    public Builder maxWaitTime(Long maxWaitTime) {
+      this.maxWaitTime = maxWaitTime;
+      return this;
+    }
+
+    /**
+     * Sets whether the query has to be dry run or not. If set, the query is not executed: if the
+     * query is valid statistics are returned on how many bytes would be processed, if the query is
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: Nit: I think this comment would sound slightly better split up into multiple sentences:

"Sets whether the query will be a dry run or not.  If set, the query is not executed.  If the query is valid, statistics are returned on how many bytes would have been processed.  If the query is invalid, an error is returned."

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+      return this;
+    }
+
+    /**
+     * Sets how long to wait for the query to complete, in milliseconds, before the request times
+     * out and returns. Note that this is only a timeout for the request, not the query. If the
+     * query takes longer to run than the timeout value, the call returns without any results and
+     * with the {@link QueryResponse#jobComplete()} set to {@code false}. You can call
+     * {@link BigQuery#getQueryResults(JobId, BigQuery.QueryResultsOption...)} to wait for the query
+     * to complete and read the results. The default value is 10000 milliseconds (10 seconds).
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: The note about the default value is nice.  I think it'd be good to add the default behavior for the other optional fields if applicable (i.e. `useQueryCache`)

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+  /**
+   * Returns the hash of the {@code QueryResponse} resource or {@code null} if not set.
+   */
+  public String etag() {
+    return etag;
+  }
+
+  /**
+   * Returns the schema of the results when the query completed successfully. Returns {@code null}
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: s/ when / if

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+import com.google.gcloud.Page;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Google Cloud BigQuery Query Response. This class contains the results of a Query Job or of a
+ * Query Request.
+ */
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Could add `@see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs/query">Query</a>` here as well. The anchor `#response` also works on that URL, so you can add it if you think it'd be helpful.

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/QueryRequestTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Test
+  public void testToBuilderIncomplete() {
+    QueryRequest queryRequest = QueryRequest.of(QUERY);
+    compareQueryRequest(queryRequest, queryRequest.toBuilder().build());
+  }
+
+  @Test
+  public void testBuilder() {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/QueryRequestTest.java]: Nit: could add a test where you try to build a QueryRequest with a null query string and expect a NPE.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+import java.util.Objects;
+
+/**
+ * Google Cloud BigQuery Query Request. This class can be used to run a BigQuery SQL query and
+ * return results if the query completes within a specified timeout. The query results are saved to
+ * a temporary table that is deleted approximately 24 hours after the query is run. Query is run
+ * through a BigQuery Job whose identity can be accessed via {@link QueryResponse#job()}.
+ *
+ * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs/query">Query</a>
+ */
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: Should we also include a link to the SQL syntax [reference](https://cloud.google.com/bigquery/query-reference)?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    private Boolean dryRun;
+    private Boolean useQueryCache;
+
+    private Builder() {}
+
+    /**
+     * Sets the BigQuery query to be executed.
+     */
+    public Builder query(String query) {
+      this.query = query;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: `checkNotNull`?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+      return this;
+    }
+
+    public QueryRequest build() {
+      return new QueryRequest(this);
+    }
+  }
+
+  private QueryRequest(Builder builder) {
+    query = checkNotNull(builder.query);
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: I think we should do the checks in the `Builder` so failure happens sooner than later (and  may miss some context).

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+      this.defaultDataset = defaultDataset;
+      return this;
+    }
+
+    /**
+     * Sets how long to wait for the query to complete, in milliseconds, before the request times
+     * out and returns. Note that this is only a timeout for the request, not the query. If the
+     * query takes longer to run than the timeout value, the call returns without any results and
+     * with the {@link QueryResponse#jobComplete()} set to {@code false}. You can call
+     * {@link BigQuery#getQueryResults(JobId, BigQuery.QueryResultsOption...)} to wait for the query
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: Are BigQuery#getQueryResults and BigQuery.QueryResultsOption coming in the next PR?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+        && Objects.equals(totalBytesProcessed, response.totalBytesProcessed)
+        && Objects.equals(jobComplete, response.jobComplete)
+        && Objects.equals(errors, response.errors)
+        && Objects.equals(cacheHit, response.cacheHit);
+  }
+
+  static Builder builder() {
+    return new Builder();
+  }
+}
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Nit, add new line.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Google Cloud BigQuery Query Response. This class contains the results of a Query Job or of a
+ * Query Request.
+ */
+public class QueryResponse implements Serializable {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: I think it would be nice to add a code snippet to show how to use this class and check for jobComplete or get the results.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+  /**
+   * Returns the query result as a paginated list of rows, if the query completed successfully.
+   * Returns {@code null} otherwise.
+   */
+  public Page<List<FieldValue>> rows() {
+    return rows;
+  }
+
+  /**
+   * Returns the total number of bytes processed for the query.
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Any reason for not including "If this query was a dry run, this is the number of bytes that would be processed if the query were run."?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  public Long totalBytesProcessed() {
+    return totalBytesProcessed;
+  }
+
+  /**
+   * Returns whether the job running the query has completed or not. If {@link #rows()} and
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: I know that this is a copy/paste from [here](https://cloud.google.com/bigquery/docs/reference/v2/jobs/query), but I think we can phrase it better. (use null instead of present) and make it clear this method should be used to check if results should be available.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  public Boolean jobComplete() {
+    return jobComplete;
+  }
+
+  /**
+   * Returns errors and warnings encountered during the running of the job, if any. Errors here do
+   * not necessarily mean that the job has completed or was unsuccessful.
+   */
+  public List<BigQueryError> errors() {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: would `executionErrors` make it more consistent?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+    return errors;
+  }
+
+  /**
+   * Returns whether the query result was fetched from the query cache.
+   *
+   * @see <a href="https://cloud.google.com/bigquery/querying-data#querycaching">Query Caching</a>
+   */
+  public Boolean cacheHit() {
+    return cacheHit;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: can this ever be null?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+    return totalBytesProcessed;
+  }
+
+  /**
+   * Returns whether the job running the query has completed or not. If {@link #rows()} and
+   * {@link #totalRows()} are present, this method will always return {@code true}. If this method
+   * returns {@code false}, {@link #totalRows()} will not be available.
+   */
+  public Boolean jobComplete() {
+    return jobComplete;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: can this ever be `null`?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+   */
+  public Page<List<FieldValue>> rows() {
+    return rows;
+  }
+
+  /**
+   * Returns the total number of bytes processed for the query.
+   */
+  public Long totalBytesProcessed() {
+    return totalBytesProcessed;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: can this ever be `null`?

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/QueryResponseTest.java [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+  private static final Long TOTAL_ROWS = 42L;
+  private static final PageImpl.NextPageFetcher<List<FieldValue>> FETCHER =
+      new PageImpl.NextPageFetcher<List<FieldValue>>() {
+        @Override
+        public Page<List<FieldValue>> nextPage() {
+          return null;
+        }
+      };
+  private static final Page<List<FieldValue>> ROWS =
+      new PageImpl<List<FieldValue>>(FETCHER, "cursor", ImmutableList.<List<FieldValue>>of());
```
Comment (aozarov) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/QueryResponseTest.java]: Nit, `new PageImpl<>`

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java (hunk 19 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Yes, this is {@code null} when the query has not completed yet.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java (hunk 17 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Yes, this is {@code null} when the query has not completed yet.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java (hunk 18 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: In the tests I run this seems to always be set. I'll use `boolean` instead.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java (hunk 11 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryRequest.java]: Yes but I better remove these links and add them later to avoid merging commits that break the build

**Comment by mziccard** [comment]
Ehi guys, thanks for the feedback. I fixed all comments and answered the open questions.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+ *
+ * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs/getQueryResults">Get Query
+ *     Results</a>
+ * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs/query">Query</a>
+ */
+public class QueryResponse implements Serializable {
+
+  private static final long serialVersionUID = 3549226764825005655L;
+
+  private final String etag;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Do you know why etag is only returned by getQueryResult and not by `query` (even if completed)?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+ *    }
+ *    List<BigQueryError> executionErrors = response.executionErrors();
+ *    Page<List<FieldValue>> rows = response.rows();
+ * }</pre>
+ *
+ * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs/getQueryResults">Get Query
+ *     Results</a>
+ * @see <a href="https://cloud.google.com/bigquery/docs/reference/v2/jobs/query">Query</a>
+ */
+public class QueryResponse implements Serializable {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: The public [docs](https://cloud.google.com/bigquery/docs/reference/v2/jobs/query) for query suggests that one could also use `bigquery.tabledata.list(table_id)` to query the temporary table that is created for the associated request. Where do you get the tableId from?

Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: I am about to ask that to the BigQuery team. The table is not part of the query response. You can still identify the table from the UI/by listing all tables but this is worth asking.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+
+  private static final long serialVersionUID = 3549226764825005655L;
+
+  private final String etag;
+  private final Schema schema;
+  private final JobId job;
+  private final Long totalRows;
+  private final Page<List<FieldValue>> rows;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: As `schema`, `totalRows`, `totalBytesProcessed`, `cacheHit`, and `rows` are only available when `jobCompleted` did you consider moving them to something like a `QueryResult` class that would extends `PageImpl` and add these accessors (`rows` would be part of the standard page)?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java (hunk 21 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: This is something else worth asking, I see no reason why this should be missing (even if not completed yet).

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java (hunk 23 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Do you think this would be more understandable/usable? I didn't want to mix non-pageable properties with pageable ones.

Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: I slightly leaning toward having a separate class that will have all values (so can use primitive) when result is available rather than having one class with all values but subset of them are not available based on a condition. 

We could create QueryResult that returns rows as a page or have that class extends Page. 

I don't feel strongly about it, so if you think current experience is better we can defer that decision.

Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: I'll give this a try now so we can see how it looks like.

**Comment by mziccard** [comment]
@ajkannan I fixed the typo in 9756dff
@aozarov I added the `QueryResult` class in ea54699

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/SerializationTest.java [hunk 24]
```diff
@@ -212,7 +233,7 @@ public void testModelAndRequests() throws Exception {
         TABLE_SCHEMA, TABLE_INFO, VIEW_INFO, EXTERNAL_TABLE_INFO, INLINE_FUNCTION, URI_FUNCTION,
         JOB_STATISTICS, EXTRACT_STATISTICS, LOAD_STATISTICS, QUERY_STATISTICS, BIGQUERY_ERROR,
         JOB_STATUS, JOB_ID, COPY_JOB, EXTRACT_JOB, LOAD_JOB, QUERY_JOB, INSERT_ALL_REQUEST,
-        INSERT_ALL_RESPONSE, FIELD_VALUE};
+        INSERT_ALL_RESPONSE, FIELD_VALUE, QUERY_REQUEST, QUERY_RESPONSE};
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/SerializationTest.java]: `QUERY_RESULT` is created above but not included in this list.

Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/SerializationTest.java]: Actually not necessary, ignore that comment

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+  private final String etag;
+  private final JobId job;
+  private final boolean jobComplete;
+  private final List<BigQueryError> executionErrors;
+
+  static final class Builder {
+
+    private QueryResult result;
+    private String etag;
+    private JobId job;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Nit, why not call it `jobId`?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+ * Query Request.
+ *
+ * <p>Example usage of a query response:
+ * <pre>    {@code
+ *    QueryResponse response = bigquery.query(request);
+ *    while (!response.jobComplete()) {
+ *      response = bigquery.getQueryResults(response.job());
+ *      Thread.sleep(1000);
+ *    }
+ *    List<BigQueryError> executionErrors = response.executionErrors();
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Nit, we should either use `executionErrors` or add a comment suggesting a possible check.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+ */
+public class QueryResponse implements Serializable {
+
+  private static final long serialVersionUID = 3549226764825005655L;
+
+  private final QueryResult result;
+  private final String etag;
+  private final JobId job;
+  private final boolean jobComplete;
+  private final List<BigQueryError> executionErrors;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: I am curious. is there an easy way for a user to tell that query completely failed (e.g. bad syntax)?
Does such validation happen before job starts (and sending the query request will fail)?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+  public boolean jobComplete() {
+    return jobComplete;
+  }
+
+  /**
+   * Returns errors and warnings encountered during the running of the job, if any. Errors here do
+   * not necessarily mean that the job has completed or was unsuccessful.
+   */
+  public List<BigQueryError> executionErrors() {
+    return executionErrors;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: can this ever be `null` or does `isEmpty` indicate no errors? My preference would be the latter (and possible add `hasErrors`).

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResult.java [hunk 29]
```diff
@@ -0,0 +1,176 @@
...
+   * Returns whether the query result was fetched from the query cache.
+   *
+   * @see <a href="https://cloud.google.com/bigquery/querying-data#querycaching">Query Caching</a>
+   */
+  public boolean cacheHit() {
+    return cacheHit;
+  }
+
+  /**
+   * Returns the schema of the results.
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResult.java]: I assume this could still be `null`? "Present only when the query completes **successfully**."

**Comment by aozarov** [comment]
Looks great! I have few more questions/nits.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java (hunk 27 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Query is validated before starting the job. The query request fails with error 400.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResult.java (hunk 29 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResult.java]: Yes

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java (hunk 28 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/QueryResponse.java]: Yes if there are no errors this is null, I can update the constructor to set this to an empty list.

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/QueryResultTest.java [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+      .pageFetcher(FETCHER)
+      .results(ImmutableList.<List<FieldValue>>of())
+      .cacheHit(CACHE_HIT)
+      .build();
+  private static final QueryResult QUERY_RESULT_INCOMPLETE = QueryResult.builder()
+      .totalBytesProcessed(TOTAL_BYTES_PROCESSED)
+      .build();
+
+  @Test
+  public void testBuilder() {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/QueryResultTest.java]: Should there be an `assertEquals` for that the `pageFetcher`?

Comment (mziccard) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/QueryResultTest.java]: I don't have access to it, this is why I am testing nextPage and nextPageCursor instead.

**Comment by mziccard** [comment]
c6e53c8 fixes latest comments and makes checkstyle happy
